### PR TITLE
Disc 92 license return non access ids

### PR DESCRIPTION
**Pull request by thomasegense**
When filtering Id's (normal ID field or resource-ID field), also return ID's that does exist, but where access was not allowed.

Here is a call to devel11 that shows the new output. 
There are 3 resource ID's in the input. 1) does not exist, 2) does exist and access, **3) no access but does exist.**

EDIT this jira, to get the correct curl command. Even code-template does not fix it. 

curl -X POST "http://devel11:10001/ds-license/v1/checkAccessForResourceIds" -H  "accept: application/json" -H  "Content-Type: application/json" -d "{\"presentationType\":\"Search\",\"accessIds\":[\"id1\",\"/DAMJP2/DAM/Samlingsbilleder/0000/750/978/DNF_1955-00298_00092\",\"/DAMJP2/DAM/Samlingsbilleder/0000/532/736/KHP0147_26_037\"],\"attributes\":[{\"attribute\":\"everybody\",\"values\":[\"yes\"]}]}"

The response.
.....
accessIds":["/DAMJP2/DAM/Samlingsbilleder/0000/750/978/DNF_1955-00298_00092"],
"nonAccessIds":["/DAMJP2/DAM/Samlingsbilleder/0000/532/736/KHP0147_26_037"]
....

### CODE REVIEW

**Review by tokee — CHANGES_REQUESTED** [review]
From a functional point of view the code seems to do as intended. The extended response is directly usable by the frontend.

The comments in this pull request are on
1. Javadoc
2. Logging
3. Code structure

And an idea at the last moment: How about adding a third list of IDs: Non-existing? That way the caller has easy access to all IDs in all possible states.

**Review by thomasegense — COMMENTED** [review]
Changes from review:
    
    1) Added list of non-existing IDs to output
    2) refactoring
    3) some javadoc
    4) refactoring retailAll method so it is done in solr method so all
    callers does have to have do it.

File: src/main/java/dk/kb/license/solr/AbstractSolrJClient.java [hunk 1]
```diff
@@ -31,72 +29,51 @@ public class AbstractSolrJClient {
...
+          query.setFilterQueries(queryPartAccess);
+          log.debug("filter:"+queryPartAccess);
+        }
         
-        query.setFields(filterResourceIdField); //only this field is used from resultset
+        query.setFields(solrIdField); //only this field is used from resultset
         query.setRows(Math.min(10000, ids.size()*200)); // Powerrating... each page can have max 200 segments (rare).. with 20 pages query this is 4000..               
         query.set("facet", "false"); //  Must be parameter set, because this java method does NOT work: query.setFacet(false);          
         QueryResponse response = solrServer.query(query); 
         log.info("response:"+response);
```
[the remark this answers is not in the document]
Comment (tokee) [reply src/main/java/dk/kb/license/solr/AbstractSolrJClient.java]: The `log.info`-thing is also prevalent in `LicenseValidator`.

File: src/main/java/dk/kb/license/validation/LicenseValidator.java [hunk 2]
```diff
@@ -25,628 +26,617 @@
...
-		//validate
-		if (input.getAttributes() == null || input.getAttributes().size() == 0){
-			log.error("No attributes defined in input.");
-			throw new IllegalArgumentException("No attributes defined in input");
-		}
+            Collections.unmodifiableList(Arrays.asList(new String[] {LOCALE_DA,LOCALE_EN}));
+
+    // The following 3 methods are the API
+
+    //TODO shitload of javadoc
```
[the remark this answers is not in the document]
Comment (tokee) [reply src/main/java/dk/kb/license/validation/LicenseValidator.java:41]: This means: Please write JavaDocs for all public methods that are `public` and non-trivial get/set/Override.

[the remark this answers is not in the document]
Comment (thomasegense) [reply src/main/java/dk/kb/license/validation/LicenseValidator.java:41]: This metode is back when module was ported and there it was agreed to spent time javadoc'ing all methods. 
Instead spending time on the open API+module documentation.

File: src/main/java/dk/kb/license/validation/LicenseValidator.java [hunk 3]
```diff
@@ -25,628 +26,617 @@
...
+
+
+    //TODO shitload of javadoc
+    public static CheckAccessForIdsOutputDto checkAccessForIds(CheckAccessForIdsInputDto input, String solrIdField) throws Exception{
+
         if  (input.getAccessIds() == null || input.getAccessIds().size() == 0){
             throw new InvalidArgumentServiceException("No ID's in input");          
         }
-        
+        log.info("checkAccessForResourceID callled for presentationtype:"+input.getPresentationType());
```
[the remark this answers is not in the document]
Comment (thomasegense) [reply src/main/java/dk/kb/license/validation/LicenseValidator.java:97]: Changed to debug, it is spammy but very useful in this phase of the project to figure out why it does not behave as expected.

File: src/main/java/dk/kb/license/validation/LicenseValidator.java [hunk 4]
```diff
@@ -25,628 +26,617 @@
...
             log.info("#filtered id for server ("+input.getPresentationType()+") "+ server.getServerUrl() +" : "+filteredIds.size());
             filteredIdsSet.addAll(filteredIds);
         }
         //Now we have to remove remove ID's not asked for that are here because of multivalue field. (set intersection)
         filteredIdsSet.retainAll(input.getAccessIds());
-                        
         output.setAccessIds(new ArrayList<String>(filteredIdsSet));
         //Sanity check!
         if (output.getAccessIds().size() > input.getAccessIds().size()){
             throw new InvalidArgumentServiceException("Security problem: More Id's in output than input. Check for query injection.");
```
[the remark this answers is not in the document]
Comment (thomasegense) [reply src/main/java/dk/kb/license/validation/LicenseValidator.java:125]: Log on warn now,  but method  also not part of this PR.

File: src/main/java/dk/kb/license/validation/LicenseValidator.java [hunk 5]
```diff
@@ -25,628 +26,617 @@
...
+
+            //Now we have to remove remove ID's not asked for that are here because of multivalue field. (set intersection)
+            noAccessfilteredIdsSet.retainAll(noAccessIdList);            
+        }
+
+        ArrayList<String> noAccessfilteredIdsList = new ArrayList<String>();
+        noAccessfilteredIdsList.addAll(noAccessfilteredIdsSet);        
+        return noAccessfilteredIdsList;
+
+    }
```
[the remark this answers is not in the document]
Comment (thomasegense) [reply src/main/java/dk/kb/license/validation/LicenseValidator.java]: Missed that comment. Is fixed now.

File: src/main/java/dk/kb/license/solr/AbstractSolrJClient.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (thomasegense) [reply src/main/java/dk/kb/license/solr/AbstractSolrJClient.java]: Did a little more clearing. I am pleased with the  ds-license.log on devel11 for now. 
Some of the spammy changed to debug

File: src/main/java/dk/kb/license/validation/LicenseValidator.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (tokee) [reply src/main/java/dk/kb/license/validation/LicenseValidator.java:41]: Grudgingly accepted.

File: src/main/java/dk/kb/license/validation/LicenseValidator.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (tokee) [reply src/main/java/dk/kb/license/validation/LicenseValidator.java]: The method is more generic now, but the code at lines 108-122 are the same as before.

If you add `queryPartAccess` as a parameter to `filterResourceIDsThatExists`, the lines 108-122 can be replaced with a single call to the method.

[the remark this answers is not in the document]
Comment (thomasegense) [reply src/main/java/dk/kb/license/validation/LicenseValidator.java]: I was totally blind to this! Is fixed now. Obvious refactoring stuff..

**Pull request merged**